UefiPayloadPkg/Library/FdtParserLib/FdtParserLib.c

```diff
@@ -0,0 +1,905 @@
+#include <Library/PrintLib.h>
+#include <Library/FdtLib.h>
+#include <Protocol/PciHostBridgeResourceAllocation.h>
+#include <Protocol/PciIo.h>
+
+typedef enum {
+  ReservedMemory = 1,
+  Memory,
+  FrameBuffer,
+  PciRootBridge,
+  Options,
+  DoNothing
+} FDT_NODE_TYPE;
+
+#define MEMORY_ATTRIBUTE_DEFAULT  (EFI_RESOURCE_ATTRIBUTE_PRESENT                   | \
+                                     EFI_RESOURCE_ATTRIBUTE_INITIALIZED             | \
+                                     EFI_RESOURCE_ATTRIBUTE_TESTED                  | \
+                                     EFI_RESOURCE_ATTRIBUTE_UNCACHEABLE             | \
+                                     EFI_RESOURCE_ATTRIBUTE_WRITE_COMBINEABLE       | \
+                                     EFI_RESOURCE_ATTRIBUTE_WRITE_THROUGH_CACHEABLE | \
+                                     EFI_RESOURCE_ATTRIBUTE_WRITE_BACK_CACHEABLE    )
+
+#define ROOT_BRIDGE_SUPPORTS_DEFAULT  (EFI_PCI_IO_ATTRIBUTE_VGA_IO_16 | \
+                              EFI_PCI_IO_ATTRIBUTE_VGA_PALETTE_IO_16 | \
+                              EFI_PCI_IO_ATTRIBUTE_ISA_IO_16 | \
+                              EFI_PCI_IO_ATTRIBUTE_IDE_PRIMARY_IO | \
+                              EFI_PCI_IO_ATTRIBUTE_VGA_IO | \
+                              EFI_PCI_IO_ATTRIBUTE_VGA_MEMORY | \
+                              EFI_PCI_IO_ATTRIBUTE_VGA_PALETTE_IO | \
+                              EFI_PCI_IO_ATTRIBUTE_ISA_IO | \
+                              EFI_PCI_IO_ATTRIBUTE_ISA_MOTHERBOARD_IO )
+
+extern VOID                         *mHobList;
+UNIVERSAL_PAYLOAD_PCI_ROOT_BRIDGES  *mPciRootBridgeInfo = NULL;
+
+/**
+  Build a Handoff Information Table HOB
+
+  This function initialize a HOB region from EfiMemoryBegin to
+  EfiMemoryTop. And EfiFreeMemoryBottom and EfiFreeMemoryTop should
+  be inside the HOB region.
+
+  @param[in] EfiMemoryBottom       Total memory start address
+  @param[in] EfiMemoryTop          Total memory end address.
+  @param[in] EfiFreeMemoryBottom   Free memory start address
+  @param[in] EfiFreeMemoryTop      Free memory end address.
+
+  @return   The pointer to the handoff HOB table.
+
+**/
+EFI_HOB_HANDOFF_INFO_TABLE *
+EFIAPI
+HobConstructor (
+  IN VOID  *EfiMemoryBottom,
+  IN VOID  *EfiMemoryTop,
+  IN VOID  *EfiFreeMemoryBottom,
+  IN VOID  *EfiFreeMemoryTop
+  );
+
+/**
+  It will check device node from FDT.
+
+  @param[in]  NodeString        Device node name string.
+  @param[in]  Depth             Check layer of Device node , only parse the 1st layer
+
+  @return FDT_NODE_TYPE         what type of the device node.
+**/
+FDT_NODE_TYPE
+CheckNodeType (
+  CHAR8  *NodeString,
+  INT32  Depth
+  )
+{
+  DEBUG ((DEBUG_INFO, "\n CheckNodeType  %a   \n", NodeString));
+  if (AsciiStrnCmp (NodeString, "reserved-memory", AsciiStrLen ("reserved-memory")) == 0 ) {
+    return ReservedMemory;
+  } else if (AsciiStrnCmp (NodeString, "memory@", AsciiStrLen ("memory@")) == 0 ) {
+    if (Depth == 1) {

```

Comment:
We should not need Depth == 1 here because if it was reserved-memory->memory, it should have been taken care by ReservedMemory sub-function.

Reply:
resolved.